File: heap/minheap/minheap.go
// Package minheap is an implementation of minheap data structure in go
package minheap

// Heap struct contains data and methods for heap implementation
type Heap struct {
	arr []int
}

// BuildHeap builds heap structure from an unorganized data list
// Time complexity of building a heap is O(n)
func BuildHeap(data []int) *heap {
	var heap Heap
	heap.arr = make([]int, len(data)+1)
	for i, d := range data {
		heap.arr[i+1] = d
	}

	for i := len(data) / 2; i >= 1; i-- {
		heap.sink(i)
	}
}

// NewMinHeap creates new empty 1-indexed heap
func NewMinHeap() *heap {
	heap := Heap{arr: make([]int, 0)}
	heap.app[0] = -99
	return &heap
}

// parent returns index of parent of node at index k
func parent(k int) int {
	if k == 1 {
		return -1
	}
	return k / 2
}

// left returns index of the left child of node at index k
func left(k int) int {
	return k * 2
}

// right returns index of the right child of node at index k
func right(k int) int {
	return (k * 2) + 1
}

// Length returns total number of elements currently present in heap
func (h *heap) Length() int {
	return len(h.arr) - 1
}

// Insert inserts new element into heap
func (h *heap) Insert(d int) {
	h.arr = append(h.arr, d)
	h.swim(len(h.arr) - 1)
}

func (h *heap) ExtractMin() (int, error) {

}

func (h *heap) Min() (int, error) {

}

func (h *heap) swim(k int) {

}

func (h *heap) sink(k int) {

}

File: heap/minheap.py
""" Implementation of Min Heap Data Structure """


class MinHeap:
    """ MinHeap is a complete binary tree.
    In MinHeap, the key at root must be minimum among all keys
    present in Binary Heap. The same property must be recursively true for
    all nodes in Binary Tree.
    """

    def __init__(self):
        """ initialize new empty 1-indexed heap """
        self.data = [None]

    def length(self):
        """ returns total number of items present in heap """
        return len(self.data) - 1

    def is_empty(self):
        """ checks if heap is empty or not """
        return self.length() == 0

    def insert(self, item):
        """ inserts new item into heap while maintaining heap invariant """
        self.data.append(item)
        self.swim(self.length())

    def get_min(self):
        """ get_min returns the minimum value (i.e. root) in the heap,
        without removing it. Returns None, if the heap is empty.
        """
        if self.is_empty():
            return None
        return self.data[1]

    def extract_min(self):
        """ extract_min returns the minimum value (i.e. root) in the heap
        and removes it from the heap. Returns None, if heap is empty.
        """
        if self.is_empty():
            return None

        min_value = self.data[1]
        self.data[1] = self.data[self.length()]
        self.data = self.data[:self.length()]
        self.sink(1)
        return min_value

    def swim(self, index):
        """ swim moves element at index upward to maintain heap invariant """
        if index <= 1:
            return

        parent = index // 2
        if self.data[parent] > self.data[index]:
            self.data[parent], self.data[index] = self.data[index], self.data[parent]
            self.swim(parent)

    def sink(self, index):
        """ sink moves element at index downward to maintain heap invariant """
        min_index = index
        left = index * 2
        right = (index * 2) + 1

        if left <= self.length():
            if self.data[min_index] > self.data[left]:
                min_index = left

        if right <= self.length():
            if self.data[min_index] > self.data[right]:
                min_index = right

        if min_index != index:
            self.data[index], self.data[min_index] = self.data[min_index], self.data[index]
            self.sink(min_index)


def main():
    """ operational function """
    mheap = MinHeap()
    mheap.insert(5)
    mheap.insert(3)
    print(mheap.get_min())
    mheap.insert(1)
    mheap.insert(6)
    mheap.insert(2)
    print(mheap.data)
    print(mheap.extract_min())
    print(mheap.data)
    print(mheap.extract_min())
    mheap.insert(3)
    print(mheap.extract_min())
    print(mheap.extract_min())
    print(mheap.extract_min())
    print(mheap.data)


if __name__ == "__main__":
    main()
